File: updraft_sdk/src/main/java/com/apsswithlove/updraft_sdk/interactor/CheckFeedbackEnabledInteractor.java
package com.apsswithlove.updraft_sdk.interactor;

import android.content.Context;
import android.content.SharedPreferences;
import com.apsswithlove.updraft_sdk.api.ApiWrapper;
import io.reactivex.Single;

public class CheckFeedbackEnabledInteractor {

    private static final String FEEDBACK_ENABLED_STORAGE = "feedback_enabled_storage";

    private static final String IS_FEEDBACK_ENABLED_PROPERTY = "is_feedback_enabled_property";

    private ApiWrapper mApiWrapper;
    private SharedPreferences mSharedPreferences;

    public CheckFeedbackEnabledInteractor(ApiWrapper apiWrapper, Context context) {
        mApiWrapper = apiWrapper;
        mSharedPreferences = context.getSharedPreferences(FEEDBACK_ENABLED_STORAGE, Context.MODE_PRIVATE);

    }

    public Single<CheckFeedbackResultModel> run() {
        return mApiWrapper.isFeedbackEnabled()
                .map(isEnabled -> {
                    boolean previouslyEnabled = mSharedPreferences.getBoolean(IS_FEEDBACK_ENABLED_PROPERTY, false);
                    boolean showAlert = true;
                    if (!isEnabled && !previouslyEnabled) {
                        showAlert = false;
                    }
                    if (isEnabled && previouslyEnabled) {
                        showAlert = false;
                    }
                    int alertType = CheckFeedbackResultModel.ALERT_TYPE_HOW_TO_GIVE_FEEDBACK;
                    if (!isEnabled && previouslyEnabled) {
                        alertType = CheckFeedbackResultModel.ALERT_TYPE_FEEDBACK_DISABLED;
                    }
                    mSharedPreferences.edit().putBoolean(IS_FEEDBACK_ENABLED_PROPERTY, isEnabled).apply();
                    return new CheckFeedbackResultModel(showAlert, alertType, isEnabled);
                });
    }
}

File: updraft-sdk/src/main/java/com/appswithlove/updraft/interactor/CheckFeedbackEnabledInteractor.kt
package com.appswithlove.updraft.interactor

import android.content.Context
import android.content.SharedPreferences
import com.appswithlove.updraft.api.ApiWrapper
import io.reactivex.Single
import androidx.core.content.edit

class CheckFeedbackEnabledInteractor(
    private val apiWrapper: ApiWrapper,
    context: Context
) {

    companion object {
        private const val FEEDBACK_ENABLED_STORAGE = "feedback_enabled_storage"
        private const val IS_FEEDBACK_ENABLED_PROPERTY = "is_feedback_enabled_property"
    }

    private val sharedPreferences: SharedPreferences =
        context.getSharedPreferences(FEEDBACK_ENABLED_STORAGE, Context.MODE_PRIVATE)

    fun run(): Single<CheckFeedbackResultModel> {
        return apiWrapper.isFeedbackEnabled()
            .map { isEnabled ->
                val previouslyEnabled = sharedPreferences.getBoolean(IS_FEEDBACK_ENABLED_PROPERTY, false)

                val showAlert = when {
                    !isEnabled && !previouslyEnabled -> false
                    isEnabled && previouslyEnabled -> false
                    else -> true
                }

                val alertType = if (!isEnabled && previouslyEnabled) {
                    CheckFeedbackResultModel.ALERT_TYPE_FEEDBACK_DISABLED
                } else {
                    CheckFeedbackResultModel.ALERT_TYPE_HOW_TO_GIVE_FEEDBACK
                }

                sharedPreferences.edit {
                    putBoolean(IS_FEEDBACK_ENABLED_PROPERTY, isEnabled)
                }

                CheckFeedbackResultModel(showAlert, alertType, isEnabled)
            }
    }
}
